Support parsing hsl() colors on the fast path

smfr opened the pull request:
#### 9599027820b8f3dce4f1d40182cb32021563202e
<pre>
Support parsing hsl() colors on the fast path
<a href="https://bugs.webkit.org/show_bug.cgi?id=276836">https://bugs.webkit.org/show_bug.cgi?id=276836</a>
<a href="https://rdar.apple.com/132110971">rdar://132110971</a>

Reviewed by Tim Nguyen.

Add support to CSSParserFastPaths for parsing `hsl()` and `hsla()` colors.

`parseSimpleAngle()` is enhanced to allow for optional units, which these colors allow.
Fix `checkForValidDouble()` to disallow a value with a trailing decimal point, since
CSS doesn&apos;t allow this.

Add some tests for invalid color parsing, to ensure we don&apos;t hit any of the `span&lt;&gt;` bounds
checks.

* LayoutTests/fast/css/invalid-color-parsing-hsl-expected.txt: Added.
* LayoutTests/fast/css/invalid-color-parsing-hsl.html: Added.
* LayoutTests/fast/css/invalid-color-parsing-rgb-expected.txt: Added.
* LayoutTests/fast/css/invalid-color-parsing-rgb.html: Added.
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::parseSimpleAngle):
(WebCore::checkForValidDouble):
(WebCore::parseRGBAlphaValue):
(WebCore::mightBeHSLA):
(WebCore::mightBeHSL):
(WebCore::parseLegacyHSL):
(WebCore::parseNumericColor):
(WebCore::parseTransformAngleArgument):
(WebCore::parseAlphaValue): Deleted.

Canonical link: <a href="https://commits.webkit.org/281160@main">https://commits.webkit.org/281160@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3d0e59a8055e549abe78c8b701ae8420d3ab588a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58951 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38280 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11443 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62593 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9395 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61081 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45931 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9597 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47674 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6690 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60982 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35815 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50979 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28533 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32540 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8270 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8399 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54492 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8550 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64284 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2864 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8517 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54992 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2874 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51006 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55095 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13055 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2398 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34109 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35193 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36278 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34939 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on LayoutTests/fast/css/invalid-color-parsing-hsl.html:
```diff
@@ -0,0 +1,91 @@
+<html>
+<head>
+    
+</head>
```
nit: useless head

nt1m commented on LayoutTests/fast/css/invalid-color-parsing-hsl.html:
```diff
@@ -0,0 +1,91 @@
+        const actualResultDiv = document.createElement('div');
+
+        for (const test of test_defs) {
+
```
nit: useless blank line

nt1m commented on LayoutTests/fast/css/invalid-color-parsing-hsl.html:
```diff
@@ -0,0 +1,91 @@
+    function test()
+    {
+        debug('Test parsing various invalid hsl colors');
+        const test_defs = [
```
```suggestion
        const tests = [
```

nt1m commented on LayoutTests/fast/css/invalid-color-parsing-rgb.html:
```diff
@@ -0,0 +1,46 @@
+    function test()
+    {
+        debug('Test parsing various invalid rgb colors');
+        const test_defs = [
```
ditto

nt1m commented on LayoutTests/fast/css/invalid-color-parsing-rgb.html:
```diff
@@ -0,0 +1,46 @@
+        const actualResultDiv = document.createElement('div');
+
+        for (const test of test_defs) {
+
```
ditto

nt1m commented on LayoutTests/fast/css/invalid-color-parsing-rgb.html:
```diff
@@ -0,0 +1,46 @@
+<html>
+<head>
+    
+</head>
```
ditto

nt1m commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -481,6 +511,106 @@ static std::optional<SRGBA<uint8_t>> parseHexColorInternal(std::span<const Chara
+    if (!saturation)
+        return std::nullopt;
+
+    // FIXME: Saturation behavior.
```
What does that mean?

nt1m commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -130,20 +132,24 @@ static inline bool parseSimpleLength(std::span<const CharacterType> characters,
 
 template <typename CharacterType>
-static inline bool parseSimpleAngle(std::span<const CharacterType> characters, CSSUnitType& unit, double& number)
+static inline bool parseSimpleAngle(std::span<const CharacterType> characters, bool requiresUnits, CSSUnitType& unit, double& number)
```
enum class?

nt1m commented on Source/WebCore/css/parser/CSSParserFastPaths.cpp:
```diff
@@ -481,6 +511,106 @@ static std::optional<SRGBA<uint8_t>> parseHexColorInternal(std::span<const Chara
+        return value;
+    };
+
+    auto parseAlpha = [&](std::span<const CharacterType>& characters) -> std::optional<double> {
```
I like defining the lambda right before their first use, makes code generally more sequential and easier to read

smfr committed:
Support parsing hsl() colors on the fast path
https://bugs.webkit.org/show_bug.cgi?id=276836
rdar://132110971

Reviewed by Tim Nguyen.

Add support to CSSParserFastPaths for parsing `hsl()` and `hsla()` colors.

`parseSimpleAngle()` is enhanced to allow for optional units, which these colors allow.
Fix `checkForValidDouble()` to disallow a value with a trailing decimal point, since
CSS doesn't allow this.

Add some tests for invalid color parsing, to ensure we don't hit any of the `span<>` bounds
checks.

* LayoutTests/fast/css/invalid-color-parsing-hsl-expected.txt: Added.
* LayoutTests/fast/css/invalid-color-parsing-hsl.html: Added.
* LayoutTests/fast/css/invalid-color-parsing-rgb-expected.txt: Added.
* LayoutTests/fast/css/invalid-color-parsing-rgb.html: Added.
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::parseSimpleAngle):
(WebCore::checkForValidDouble):
(WebCore::parseRGBAlphaValue):
(WebCore::mightBeHSLA):
(WebCore::mightBeHSL):
(WebCore::parseLegacyHSL):
(WebCore::parseNumericColor):
(WebCore::parseTransformAngleArgument):
(WebCore::parseAlphaValue): Deleted.

Canonical link: https://commits.webkit.org/281160@main
